Версия 3.0 ТЗ 12

Akmula opened the pull request.

Akmula committed:
Версия 3.0
ТЗ 12

Akmula committed:
Версия 3.0
ТЗ 12

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -31,10 +29,15 @@ public ErrorResponse handleUserNotFoundException(final UserNotFoundException e)
     @ExceptionHandler
     @ResponseStatus(HttpStatus.NOT_FOUND)
     public ErrorResponse handleNoResourceException(final NoResourceFoundException e) {
-        log.error("Ошибка в запросе: Ресурс '/{}' не найден!", e.getResourcePath());
```
Логирование в обработчике ошибок лучше оставить. Логи помогут разбираться с возникшими ошибками.

Akmula replied:
Вернул логирование.

Akmula replied:
Логирование вернул

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -18,62 +18,54 @@
     }
 
     @GetMapping("/{filmId}")
-    public ResponseEntity<Film> getFilmById(@PathVariable Integer filmId) {
```
Можно использовать примитивные типы `@PathVariable int filmId`

Akmula replied:
Исправил

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -18,62 +18,54 @@
 
     @GetMapping("/{filmId}")
-    public ResponseEntity<Film> getFilmById(@PathVariable Integer filmId) {
+    public ResponseEntity<FilmDto> getFilmById(@PathVariable Integer filmId) {
```
Можно более кооротко без ResponseEntity
```
public FilmDto getFilmById(@PathVariable int filmId) {
   ...
   return filmService.getFilmById(filmId);
}
```

Akmula replied:
ResponseEntity убрал

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
Убираем из проекта, больше не понадобится

Akmula replied:
убрал

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
Убираем из проекта, больше не понадобится

Akmula replied:
убрал

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -1,105 +1,163 @@
+    public Collection<UserDto> getAllUsers() {
+        log.info("UserService - Получение всех пользователей");
+        Collection<UserDto> usersDto = UserMapper.mapToUserDtoList(userStorage.getAllUsers());
+        usersDto.forEach(userDto -> userDto.setFriends(friendshipStorage.getFriendsIds(userDto.getId())));
```
Нет требований для каждого юзера отдавать коллекцию с друзьями. Для друзей есть отдельный эндпоинт и его достаточно.
Не оптимально в цикле дергать БД по обному запросу. Лучше убрать. Или же если хочешь оставить у  пользователей коллкцию с друзьями, то подумай как обойтись только двумя запросами, первым получаем всех пользователей, вторым всех друзей, затем соединяем их.

Akmula replied:
Исправил, надеюсь норм.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/dal/BaseRepository.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Optional;
+
+@RequiredArgsConstructor
+public class BaseRepository<T> {
```
Отлично, что выделил повторяющуюся логику в базовый класс

Akmula replied:
Спасибо

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -0,0 +1,136 @@
+        );
+        film.setId(id);
+        if (film.getGenres() != null) {
+            for (Genre genre : film.getGenres()) {
```
В цикле обращаться к БД не оптимально по производительности.
Лучше для вставки коллекции жанров воспользоваться пакетной операцией jdbcTemplate.batchUpdate()
https://habr.com/ru/articles/703828/

Akmula replied:
Исправил

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -0,0 +1,136 @@
+    private static final String GET_ALL_FILM_QUERY = """
+            SELECT f.*, m.name AS mpa_name, m.description AS mpa_description
+            FROM FILMS AS f
+            JOIN MPA AS m ON f.mpa_id = m.mpa_id
```
Отлично, что заджойнил MPA

Akmula replied:
Спасибо.

VadimZharkov reviewed (changes requested):
Хорошая работа, можно улучшить

Akmula committed:
Версия 3.1
ТЗ 12
1. Вернуть логирование в обработчик ошибок.
2. Убрать в контроллерах ResponseEntity
3. В контроллерах использовать примитивные типы @PathVariable int filmId
4. Убрать InMemoryFilmStorage и InMemoryUserStorage
5. Изменить метод добавления списка друзей у пользователей, при получении всех пользователей
6. Для вставки коллекции жанров воспользоваться пакетной операцией jdbcTemplate.batchUpdate()
